File: src/main/kotlin/frc/robot/Robot.kt
package frc.robot

import edu.wpi.first.wpilibj.TimedRobot
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.CommandScheduler

object Robot : TimedRobot() {
	private lateinit var robotContainer: RobotContainer
	private var scheduledAuto: Command? = null

	override fun robotInit() {
		robotContainer = RobotContainer()
	}

	override fun robotPeriodic() {
		CommandScheduler.getInstance().run()
	}

	override fun autonomousInit() {
		scheduledAuto = robotContainer.auto?.let {
			it.schedule()
			it
		}
	}

	override fun autonomousPeriodic() {}

	override fun teleopInit() {
		scheduledAuto = scheduledAuto?.let {
			it.cancel()
			null
		}
	}

	override fun teleopPeriodic() {}

	override fun testInit() {
		CommandScheduler.getInstance().cancelAll()
	}

	override fun testPeriodic() {}
	override fun simulationPeriodic() {}
	override fun disabledPeriodic() {}
}

File: src/main/scala/frc/robot/Robot.scala
package frc.robot

import edu.wpi.first.wpilibj.TimedRobot
import edu.wpi.first.wpilibj2.command.{Command, CommandScheduler}

object Robot extends TimedRobot:
	private lazy val robotContainer = RobotContainer()
	private var scheduledAuto: Option[Command] = None

	override def robotInit(): Unit =
		robotContainer // Force evaluation

	override def robotPeriodic(): Unit =
		CommandScheduler.getInstance().run()

	override def autonomousInit(): Unit =
		scheduledAuto = robotContainer.auto.map(auto =>
			auto.schedule()
			auto
		)

	override def autonomousPeriodic(): Unit = ()

	override def teleopInit(): Unit =
		scheduledAuto = scheduledAuto.flatMap(auto =>
			auto.cancel()
			None
		)

	override def teleopPeriodic(): Unit = ()

	override def testInit(): Unit =
		CommandScheduler.getInstance().cancelAll()

	override def testPeriodic(): Unit = ()

	override def disabledPeriodic(): Unit = ()

	override def simulationPeriodic(): Unit = ()
